### PR TITLE
Fixes #29156: Failing sensitive variables values are not logged

### LINKED ISSUE #29156: Sensitive values are sent to stderr if there are type errors
If you define a sensitive variable with a type definition then try to assign a value to it with an invalid type, that value is printed out verbatim.  Terraform is usually very good at not printing sensitive variables so it's possible this is intended behaviour.  However, invalid type definitions or assignments aren't uncommon when making changes so this behaviour could easily leak secrets, hence I've categorised this as a bug. Thanks!

### Terraform Version

```
1.0.1, 0.15.5
```

### Terraform Configuration Files
https://github.com/memriseed/sensitive-vars-example

### Expected Behavior
If a sensitive variable encounters an issue with an invalid value, it should redact or suppress output containing that value.

### Actual Behavior
Terraform prints the following error message, containing the un-redacted contents of the values file:

```
╷
│ Error: Invalid value for input variable
│ 
│   on secrets.auto.tfvars line 5:
│    5: secret_var = ["test"]
│ 
│ The given value is not valid for variable "secret_var": string required.
╵
```

### Steps to Reproduce

1. Clone the repo above
2. `terraform init`
3. `terraform plan`

### References
I couldn't find any references to this in existing issues but there are a lot of them and the Github search tool isn't perfect - please let me know if this has already been raised!

### PR DESCRIPTION
**Pull request by gbataille**

### CODE REVIEW

**Comment by gbataille** [comment]
#29156 looked good as a first issue for me. And I agree very much with the problem stated.

**Comment by gbataille** [comment]
by the way, 

before 
Screenshot 2022-02-19 at 09 11 31


After
After

**Comment by crw** [comment]
Hi @gbataille, thanks for the contribution! We are looking into it, but I cannot commit to when this can be reviewed.

File: internal/terraform/eval_variable.go [hunk 1]
```diff
@@ -90,27 +90,37 @@ func prepareFinalInputVariableValue(addr addrs.AbsInputVariableInstance, raw *In
...
+			)
+			subject = sourceRange.ToHCL().Ptr()
+		}
+
+		// In some workflows, the operator running terraform does not have access to the variables
+		// themselves. They are for example stored in encrypted files that will be used by the CI toolset
+		// and not by the operator directly. In such a case, the failing secret value should not be
+		// displayed to the operator
+		if cfg.Sensitive {
+			subject = nil
```
Comment (alisdair) [inline internal/terraform/eval_variable.go:114]: I don't think there's any need to remove the subject here when it comes from a non-file source, and we probably also need to treat the logic for appending to the diagnostic detail message separately too. I think therefore we probably want an `if cfg.Sensitive { … }` block inside both branches of the `if nonFileSource != "" { … }` statement.

**Review by alisdair — CHANGES_REQUESTED** [review]
Thanks for working on this!

I think the general idea here is a good solution to this problem. There are a couple of improvements I'd like to make before going ahead.

Whenever possible, we should leave in the non-file source information (e.g. `set using -var="foo=..."`). I left an inline note about this.

For sensitive variables set from a file, I agree that the best option we have is to replace the source range. However, we need to inform the user where the variable value comes from, by including the .tfvars filename and line number in the detail string.

In this case, we can also follow the existing pattern for non-file sourced values of using the variable declaration as the source range, to give the user extra context.

Here are some example diagnostic designs based on these thoughts:

```
╷
│ Error: Invalid value for input variable
│
│   on main.tf line 1:
│    1: variable "foo" {
│
│ Unsuitable value for var.foo set using -var="foo=...": a bool is required.
╵
```
```
╷
│ Error: Invalid value for input variable
│
│   on main.tf line 1:
│    1: variable "foo" {
│
│ The given value is not suitable for var.foo, which is sensitive: a bool is required. Value
| declared at terraform.tfvars line 1.
╵
```

(Note the trailing full stop at the end of the description string, which is Terraform house style.)

Diagnostics for non-sensitive variable values should not change, so I've omitted these.

Let me know if anything is unclear here, or if you'd like me to review alternative designs. Thanks again!

File: internal/terraform/eval_variable.go (hunk 1 above)
Comment (gbataille) [reply internal/terraform/eval_variable.go:114]: that's how I coded it originally and changed it back to that.

But I read all your comments and they make sense to me. I'll try to take some time this week to update the PR

**Comment by gbataille** [comment]
sorry for not having come back sooner. Did not take the time.
So I agree with your comments
- the `var=` message is already such that it does not show the variable value, so no reason to change it.
- for the error coming from a variable file, I agree that it sounds better to have some snippets. I have however looked at how those "errors" are brough up, as `Diagnostics` (and actually multiple `Diagnostics` struct from different packages), all the way up.

--> I did not find yet a "smart" way to do it.
- we could say that subject is not an `hcl.Range` anymore but a `Range` + a sensitive information. If `sensitive` is true, then we display location but not snippet when rendering a `Range`. 
- we could bring a `sensitive` flag in the `Diagnostic`(s) struct. In such a case, we would display source location information but not source snippet.
Both of those are quite involved and I don't want to start on that without your opinion. Any thoughts, advices, options I failed to see?

I have in the meantime added the full stop at the end of the message and moved the logic to just one branch of the code

File: internal/terraform/eval_variable.go [hunk 2]
```diff
@@ -90,27 +90,37 @@ func prepareFinalInputVariableValue(addr addrs.AbsInputVariableInstance, raw *In
...
+			)
+			subject = sourceRange.ToHCL().Ptr()
+
+			// In some workflows, the operator running terraform does not have access to the variables
+			// themselves. They are for example stored in encrypted files that will be used by the CI toolset
+			// and not by the operator directly. In such a case, the failing secret value should not be
+			// displayed to the operator
+			if cfg.Sensitive {
+				subject = nil
+				detail += fmt.Sprintf("\n\n%s is marked as sensitive. The failing variable assignment is not printed.", addr)
```
Comment (alisdair) [inline internal/terraform/eval_variable.go:114]:
```suggestion
				detail += fmt.Sprintf("\n\n%s is marked as sensitive. Invalid value defined at %s.", addr, sourceRange.ToHCL())
```

**Review by alisdair — CHANGES_REQUESTED** [review]
Thanks for the update! I think this is really close.

While I like your ideas about adding sensitive value redaction to diagnostics, I agree that it would be a lot of work. What I'd propose for this PR is to adjust one diagnostic to include a reference to the origin of the invalid variable value in the detail text, while keeping the code snippet pointing at the variable definition. See inline suggestion.

There are also some failing tests to fix up, but once those two things are done I think this is good to go!

File: internal/terraform/eval_variable.go (hunk 2 above)
Comment (gbataille) [reply internal/terraform/eval_variable.go:114]: ok I see. feels a bit hackish in the sense that diagnostics and subject handling are managed in common way up there, but I definitely can make that work (don't know when though, sorry :| )

**Comment by gbataille** [comment]
I feel stupid for not even thinking about the "local" solution. It is much simpler, although indeed a bit redundant.

Here is the result

image

**Comment by alisdair** [comment]
I pushed a tiny tweak to add back the config declaration as the source parameter in both cases, resulting in diags like this:

diags

With that in place this looks good to me. We're in the freeze for 1.2.0 at the moment, but I'm marking this approved and plan to merge it once that freeze is over. Thanks again for working on this!

**Review by alisdair — APPROVED** [review]

**Comment by alisdair** [comment]
This change will go out in the upcoming 1.2.2 release. Thanks once more!

**Comment by gbataille** [comment]
cool, cheers

---
Gregory Bataille

**Pull request merged**